### PR TITLE
Implement React Native framework-less to storybook-react-native app

### PR DESCRIPTION
**Pull request by brianacnguyen**
## **Description**
This PR implements React Native framework-less to storybook-react-native app

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-design-system/issues#workspaces/design-system-61e8a2ae77c3a60012e5003c/issues/zh/809

## **Manual testing steps**

1. Run `yarn install`
2. Go to `/apps/storybook-react-native/ios/` ==> Run `pod install`
3. From root, run `yarn workspace @metamask/design-system-react-native start`
4. Go to `/apps/storybook-react-native/ios/` ==> Open `MetamaskStorybookReactNative.xcworkspace` with Xcode ==> Press Play

### **After**
Simulator Screenshot - iPhone 16 Pro - 2024-10-21 at 14 15 01

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs)
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: packages/design-system-react-native/LICENSE
Comment (georgewrmarshall) [inline packages/design-system-react-native/LICENSE:1]: We need this for packages

File: packages/design-system-react-native/README.md
Comment (georgewrmarshall) [inline packages/design-system-react-native/README.md:1]: We need this for packages

File: packages/design-system-react-native/package.json
Comment (georgewrmarshall) [inline packages/design-system-react-native/package.json:1]: We need this for packages

File: apps/storybook-react-native/ios/MetamaskStorybookReactNative.xcodeproj/xcshareddata/xcschemes/MetamaskStorybookReactNative.xcscheme
Comment (Cal-L) [inline apps/storybook-react-native/ios/MetamaskStorybookReactNative.xcodeproj/xcshareddata/xcschemes/MetamaskStorybookReactNative.xcscheme:1]: Confirm if this file should be commited

File: .gitignore [hunk 1]
```diff
@@ -18,6 +21,9 @@ preview-build-message.txt
 packages/*/coverage
 packages/*/dist
 packages/*/docs
+apps/*/coverage
+apps/*/dist
+apps/*/docs
```
Comment (georgewrmarshall) [inline .gitignore]: Do these folders exist in apps? I don't see them in storybook-react-native can they be removed?

If you run `find apps -type d | grep -E '/(coverage|dist|docs)$'` in the terminal there are no results. 

Screenshot 2024-10-22 at 1 01 35 PM

If you run `find apps -type d | grep -E '/(coverage|dist|android)$'` it will pick up the android directory

Screenshot 2024-10-22 at 1 01 50 PM


```suggestion
```

File: .gitignore [hunk 2]
```diff
@@ -3,13 +3,16 @@ npm-debug.log
 yarn-error.log
 node_modules
 package-lock.json
+**/package-lock.json
+**/yarn.lock
```
Comment (georgewrmarshall) [inline .gitignore]: We can remove some redundant entries in the .gitignore file to make it cleaner and easier to maintain:
1. `package-lock.json` is already covered by the line above it. The `**/` prefix is unnecessary since the single entry will ignore this file in all directories.
2. `**/yarn.lock` is not needed if we're using Yarn v3, as the Yarn v3 section later in the file already handles ignoring the necessary Yarn-related files.

```suggestion
```

File: .gitignore [hunk 3]
```diff
@@ -32,5 +38,42 @@ scripts/coverage
...
+**/ios/DerivedData/
+**/ios/.xcuserdata/
+**/ios/.xcode.env.local
+*.mode1v3
+*.pbxuser
+*.xccheckout
+*.xcuserstate
+*.xcworkspace
+!ios/Podfile.lock
+!ios/Pods/
```
Comment (georgewrmarshall) [inline .gitignore:66]: Don't we typically want to ignore `Pods` files? See mobile https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/blob/615267ea0b44bc3be657b84fbc9ff7c86688fa58/.gitignore#L16 @Cal-L could you please check these and what is appropriate to ignore?

File: .prettierignore [hunk 4]
```diff
@@ -0,0 +1,5 @@
+node_modules/
+
+# Ignore all iOS and Android folders across packages
+apps/**/ios/**
+apps/**/android/**
```
Comment (georgewrmarshall) [inline .prettierignore:5]: nit: Add line break here
```suggestion
apps/**/android/**

```

File: README.md [hunk 5]
```diff
@@ -22,7 +21,7 @@ Or, in graph form [^fn1]:
 %%{ init: { 'flowchart': { 'curve': 'bumpX' } } }%%
 graph LR;
 linkStyle default opacity:0.5
-  design_system_react_native(["@metamask/design-system-react-native"]);
+  design_system_react_native(["@metamask/storybook-react-native"]);
```
Comment (georgewrmarshall) [inline README.md]: We should keep the reference to `packages/design-system-react-native` because it's an exported module from the monorepo. However, we can remove the reference to `react-native storybook` as it's not a package. Revert the changes to this file

``` git checkout main -- README.md```

File: apps/storybook-react-native/.eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    {
+      files: ['*.js', '*.jsx'],
+      parserOptions: {
+        ecmaVersion: 2020,
+        sourceType: 'module', // Allows ES6 import/export syntax
+      },
+      rules: {
+        // Add any package-specific rules here
+      },
+    },
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/.eslintrc.js]: Because this is a new TypeScript-based React Native project, can simplify this ESLint configuration by removing the JavaScript/JSX overrides. Since we're not using any legacy JS code, we can focus solely on TS and TSX files?
```suggestion
```

File: apps/storybook-react-native/CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Changelog
+
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
+and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
+
+## [Unreleased]
+
+[Unreleased]: https://github.com/MetaMask/metamask-design-system/
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/CHANGELOG.md]: This changelog file can be removed because we aren't publishing this module

File: apps/storybook-react-native/jest.config.js [hunk 8]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable import/unambiguous */
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/jest.config.js]: We can likely remove this file as we won't be writing jest tests for the storybook app

File: apps/storybook-react-native/package.json [hunk 9]
```diff
@@ -3,7 +3,86 @@
...
+        "default": "./dist/index.cjs"
+      }
+    },
+    "./package.json": "./package.json"
+  },
+  "main": "./dist/index.cjs",
+  "types": "./dist/index.d.cts",
+  "files": [
+    "dist/"
+  ],
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/package.json]: Not sure why we are adding this back? It's mostly used for the packages

File: apps/storybook-react-native/package.json [hunk 10]
```diff
@@ -3,7 +3,86 @@
...
+    "postinstall": "cd ios && pod install || exit 0",
+    "publish:preview": "yarn npm publish --tag preview",
+    "run-android": "react-native run-android",
+    "run-ios": "react-native run-ios --scheme MetamaskDesignSystemApp",
+    "since-latest-release": "../../scripts/since-latest-release.sh",
+    "start": "react-native start",
+    "test": "NODE_OPTIONS=--experimental-vm-modules jest --reporters=jest-silent-reporter",
+    "test:clean": "NODE_OPTIONS=--experimental-vm-modules jest --clearCache",
+    "test:verbose": "NODE_OPTIONS=--experimental-vm-modules jest --verbose",
+    "test:watch": "NODE_OPTIONS=--experimental-vm-modules jest --watch"
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/package.json]: A lot of unnecessary scripts in here that can be removed

File: apps/storybook-react-native/package.json [hunk 11]
```diff
@@ -3,7 +3,86 @@
...
+    "deepmerge": "^4.2.2",
+    "eslint": "^8.44.0",
+    "eslint-plugin-ft-flow": "^3.0.11",
+    "jest": "^27.5.1",
+    "metro-react-native-babel-preset": "^0.76.0",
+    "prettier": "^2.7.1",
+    "react-test-renderer": "18.2.0",
+    "ts-jest": "^27.1.4",
+    "typedoc": "^0.24.8",
+    "typedoc-plugin-missing-exports": "^2.0.0",
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/package.json]: Probably some dependencies related to jest and other package specific scripts that can be removed. I'll create a commit

File: apps/storybook-react-native/tsconfig.build.json [hunk 12]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (georgewrmarshall) [inline apps/storybook-react-native/tsconfig.build.json]: Not sure if we need this file

File: package.json [hunk 13]
```diff
@@ -20,12 +20,13 @@
...
     "changelog:validate": "yarn workspaces foreach --all --no-private --parallel --interlaced --verbose run changelog:validate",
     "create-package": "ts-node scripts/create-package",
+    "postinstall": "cd apps/storybook-react-native/ios && pod install || exit 0",
     "lint": "yarn lint:eslint && yarn lint:misc --check && yarn constraints && yarn lint:dependencies",
     "lint:dependencies": "depcheck && yarn dedupe --check",
     "lint:dependencies:fix": "depcheck && yarn dedupe",
     "lint:eslint": "eslint . --cache --ext js,cjs,mjs,ts",
     "lint:fix": "yarn lint:eslint --fix && yarn lint:misc --write && yarn constraints --fix && yarn lint:dependencies:fix",
-    "lint:misc": "prettier '**/*.json' '**/*.md' '!**/CHANGELOG.old.md' '**/*.yml' '!.yarnrc.yml' '!merged-packages/**' --ignore-path .gitignore",
+    "lint:misc": "prettier --check 'packages/**/*.{json,md,yml}' --ignore-path .prettierignore",
```
Comment (georgewrmarshall) [inline package.json:29]: Can you give a bit more context here?

File: package.json [hunk 14]
```diff
@@ -62,6 +63,7 @@
     "@types/jest": "^27.4.1",
     "@types/lodash": "^4.14.191",
     "@types/node": "^16.18.54",
+    "@types/react-native": "^0.73.0",
```
Comment (georgewrmarshall) [inline package.json]: Should this be inside `apps/storybook-react-native/package.json`?

File: package.json [hunk 15]
```diff
@@ -20,12 +20,13 @@
     "changelog:update": "yarn workspaces foreach --all --no-private --parallel --interlaced --verbose run changelog:update",
     "changelog:validate": "yarn workspaces foreach --all --no-private --parallel --interlaced --verbose run changelog:validate",
     "create-package": "ts-node scripts/create-package",
+    "postinstall": "cd apps/storybook-react-native/ios && pod install || exit 0",
```
Comment (georgewrmarshall) [inline package.json:23]: Shout this be named `postinstall` or `podinstall`?

File: packages/design-system-tailwind-preset/package.json [hunk 16]
```diff
@@ -49,6 +49,7 @@
   "devDependencies": {
     "@metamask/auto-changelog": "^3.4.4",
     "@types/jest": "^27.4.1",
+    "@types/node": "^16.18.54",
```
Comment (georgewrmarshall) [inline packages/design-system-tailwind-preset/package.json]: Why is this necessary?

File: tsconfig.build.json [hunk 17]
```diff
@@ -1,5 +1,6 @@
 {
   "references": [
+    { "path": "./apps/storybook-react-native/tsconfig.build.json" },
```
Comment (georgewrmarshall) [inline tsconfig.build.json]: Shouldn't be needed?
```suggestion
```

File: packages/design-system-tailwind-preset/tsconfig.build.json [hunk 18]
```diff
@@ -3,7 +3,8 @@
   "compilerOptions": {
     "baseUrl": "./",
     "outDir": "./dist",
-    "rootDir": "./src"
+    "rootDir": "./src",
+    "types": ["node"]
```
Comment (georgewrmarshall) [inline packages/design-system-tailwind-preset/tsconfig.build.json]: Why is this necesary?

File: packages/design-system-tailwind-preset/tsconfig.json [hunk 19]
```diff
@@ -1,7 +1,8 @@
 {
   "extends": "../../tsconfig.packages.json",
   "compilerOptions": {
-    "baseUrl": "./"
+    "baseUrl": "./",
+    "types": ["node"]
```
Comment (georgewrmarshall) [inline packages/design-system-tailwind-preset/tsconfig.json]: Why is this necessary?

**Review by georgewrmarshall — COMMENTED** [review]
Getting there! Left a few comments. I'll push some changes which should help with the failing tests.

**Comment by georgewrmarshall** [comment]
All dependency warnings are supply chain risk which Lavamoat protects us against @SocketSecurity ignore-all

Screenshot 2024-10-22 at 3 52 00 PM
Screenshot 2024-10-22 at 3 52 09 PM
Screenshot 2024-10-22 at 3 52 34 PM
Screenshot 2024-10-22 at 3 52 41 PM

File: tsconfig.packages.build.json [hunk 20]
```diff
@@ -5,7 +5,12 @@
     "declarationMap": true,
     "emitDeclarationOnly": true,
     "inlineSources": true,
-    "sourceMap": true
+    "sourceMap": true,
+    // Prevent automatic inclusion of @types packages by using an empty array
+    // This gives us more control over type definitions and helps avoid conflicts
+    // Types must be explicitly imported in files where they are needed
+    // or included in extending tsconfig files for specific parts of the project
```
Comment (georgewrmarshall) [inline tsconfig.packages.build.json:12]: Forgot you can't add comments to json

File: tsconfig.packages.build.jsonc [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+  "compilerOptions": {
+    "declaration": true,
+    "declarationMap": true,
+    "emitDeclarationOnly": true,
+    "inlineSources": true,
+    "sourceMap": true,
+    // Prevent automatic inclusion of @types packages by using an empty array
+    // This gives us more control over type definitions and helps avoid conflicts
+    // Types must be explicitly imported in files where they are needed
+    // or included in extending tsconfig files for specific parts of the project
```
Comment (georgewrmarshall) [inline tsconfig.packages.build.jsonc:12]: JSON typically doesn't allow comments. Updating `tsconfig.packages.build.json` to `jsonc` to allow for critical comments about our typescript configuration

File: package.json [hunk 22]
```diff
@@ -20,12 +20,13 @@
...
     "lint": "yarn lint:eslint && yarn lint:misc --check && yarn constraints && yarn lint:dependencies",
     "lint:dependencies": "depcheck && yarn dedupe --check",
     "lint:dependencies:fix": "depcheck && yarn dedupe",
     "lint:eslint": "eslint . --cache --ext js,cjs,mjs,ts",
     "lint:fix": "yarn lint:eslint --fix && yarn lint:misc --write && yarn constraints --fix && yarn lint:dependencies:fix",
-    "lint:misc": "prettier '**/*.json' '**/*.md' '!**/CHANGELOG.old.md' '**/*.yml' '!.yarnrc.yml' '!merged-packages/**' --ignore-path .gitignore",
+    "lint:misc": "prettier --check 'packages/**/*.{json,jsonc,md,yml}' --ignore-path .prettierignore",
```
Comment (georgewrmarshall) [inline package.json:29]: Not sure if we want to lint some of the files in the `apps/` directory but great catch there are some left over references from core here that needed updating

File: package.json [hunk 23]
```diff
@@ -95,12 +96,16 @@
   "engines": {
     "node": "^18.18 || >=20"
   },
+  "installConfig": {
+    "hoistingLimits": "workspaces"
+  },
```
Comment (georgewrmarshall) [inline package.json:101]: Can you explain what this does?

File: package.json [hunk 24]
```diff
@@ -95,12 +96,16 @@
...
   "lavamoat": {
     "allowScripts": {
       "@lavamoat/preinstall-always-fail": false,
       "@keystonehq/bc-ur-registry-eth>hdkey>secp256k1": true,
       "babel-runtime>core-js": false,
-      "simple-git-hooks": false
+      "simple-git-hooks": false,
+      "$root$": false
```
Comment (georgewrmarshall) [inline package.json:108]: Can you explain what this is for?

File: packages/design-system-tailwind-preset/package.json (hunk 16 above)
Comment (georgewrmarshall) [reply packages/design-system-tailwind-preset/package.json:52]: This is necessary because by setting the tsconfig.packages.json types array to empty we are essentially ignoring all automatic types which included `@types/node` so now we have to set it for each individual package. We may want to revisit how our tsconfigs are setup

File: packages/design-system-tailwind-preset/tsconfig.build.json (hunk 18 above)
Comment (georgewrmarshall) [reply packages/design-system-tailwind-preset/tsconfig.build.json:7]: See https://github.com/MetaMask/metamask-design-system/pull/60/files#r1811558493

File: packages/design-system-tailwind-preset/tsconfig.json [hunk 25]
```diff
@@ -1,7 +1,8 @@
 {
   "extends": "../../tsconfig.packages.json",
   "compilerOptions": {
-    "baseUrl": "./"
+    "baseUrl": "./",
+    "types": ["node", "jest"]
```
Comment (georgewrmarshall) [inline packages/design-system-tailwind-preset/tsconfig.json:5]: See https://github.com/MetaMask/metamask-design-system/pull/60/files#r1811558493

File: package.json (hunk 13 above)
Comment (brianacnguyen) [reply package.json]: the current lint listens to gitignore, which can have unintended effects. Ive added a separate prettierignore

File: package.json (hunk 15 above)
Comment (brianacnguyen) [reply package.json:23]: postinstall

File: packages/design-system-tailwind-preset/package.json (hunk 16 above)
Comment (brianacnguyen) [reply packages/design-system-tailwind-preset/package.json:52]: This was to resolve a test failure

File: package.json (hunk 24 above)
Comment (brianacnguyen) [reply package.json:108]: Screenshot 2024-10-22 at 5 53 21 PM

File: package.json (hunk 15 above)
Comment (georgewrmarshall) [reply package.json:23]: I think the naming of this script is causing the lavamoat issue in [this comment](https://github.com/MetaMask/metamask-design-system/pull/60/files#r1811609620). When I rename this script I no longer get this Lavamoat issue. 


https://github.com/user-attachments/assets/bb0a91ed-5c68-4b9e-9fbe-b730b7ab2077

Why Not Set `"$root$": false`?

While it might be tempting to set `"$root$": false` in the Lavamoat configuration to quickly bypass the issue, doing so is **not recommended**. Here's why:

- **Disallows All Root Scripts:** Setting `"$root$": false` effectively prevents any scripts from running at the root level of the project. This blanket restriction can inadvertently block legitimate and necessary scripts, including `postinstall`, leading to further setup and functionality issues.

- **Overly Restrictive:** Rather than selectively permissive, this approach is too broad and doesn't allow for the flexibility needed to run specific scripts that are essential for the project's setup and maintenance.

- **Security Implications:** Disabling scripts at the root level entirely may lead to scenarios where necessary scripts are not executed, potentially breaking the build or deployment processes. It also doesn't provide a secure way to manage which scripts are allowed, reducing the overall effectiveness of Lavamoat's security measures.

Comment (georgewrmarshall) [reply package.json:23]: `yarn pod:install` still works

Screenshot 2024-10-23 at 12 03 08 PM

File: package.json (hunk 24 above)
Comment (georgewrmarshall) [reply package.json:105]: I don't think we want to set  `"$root$": false` here  https://github.com/MetaMask/metamask-design-system/pull/60/files#r1813382276

**Review by georgewrmarshall — COMMENTED** [review]
Left a couple comments which I can update to get this across the line. Nice work 💯

File: packages/design-system-tailwind-preset/tsconfig.json (hunk 25 above)
Comment (georgewrmarshall) [reply packages/design-system-tailwind-preset/tsconfig.json:5]: Updated here https://github.com/MetaMask/metamask-design-system/pull/60/commits/e8d4b6a8de84b19a93639877036aabf2b45dc130

File: package.json (hunk 24 above)
Comment (georgewrmarshall) [reply package.json]: Updated here https://github.com/MetaMask/metamask-design-system/pull/60/commits/e8d4b6a8de84b19a93639877036aabf2b45dc130

**Review by georgewrmarshall — DISMISSED** [review]
LGTM! Nice work 💯

File: .gitignore (hunk 3 above)
Comment (Cal-L) [reply .gitignore]: Good catch. Pods should be ignored as well

File: .gitignore [hunk 26]
```diff
@@ -32,5 +33,41 @@ scripts/coverage
...
 # typescript
 packages/*/*.tsbuildinfo
+apps/*/*.tsbuildinfo
 
+# Logs
 *storybook.log
+
+# Linux system files
+*~
+
```
Comment (Cal-L) [inline .gitignore:43]: Ignore VScode settings
```suggestion
# editor
.vscode
```

**Review by Cal-L — COMMENTED** [review]
Left some comments

File: .gitignore (hunk 26 above)
[the remark this answers is not in the document]
Comment (georgewrmarshall) [reply .gitignore]: Added!

Comment (georgewrmarshall) [reply .gitignore]: I think we want to include the `.vscode` settings it's something that @garrettbear added to add some color to our repo.

**Review by georgewrmarshall — APPROVED** [review]
